pkg/sql/schemachanger/scbuild/scbuild_test: TestBuildIsMemoryMonitored failed [generate_test_objects became slower]

pkg/sql/schemachanger/scbuild/scbuild_test.TestBuildIsMemoryMonitored [failed](https://tanzanite.cluster.engflow.com/invocations/default/6a723732-b1cb-4f15-85f0-4a6c68bc048d?testReportRun=20&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvc3FsL3NjaGVtYWNoYW5nZXIvc2NidWlsZDpzY2J1aWxkX3Rlc3Q=) on release-25.1 @ [6123ebaa10c3847899aa81b09ec4cb8d08c0bd6d](https://github.com/cockroachdb/cockroach/commits/6123ebaa10c3847899aa81b09ec4cb8d08c0bd6d):

Fatal error:

```
panic: test timed out after 15m0s
running tests:
	TestBuildIsMemoryMonitored (13m58s)

```
Stack: 

```
goroutine 273901 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x385
created by time.goFunc
	GOROOT/src/time/sleep.go:177 +0x2d
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestBuildIsMemoryMonitored
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestBuildIsMemoryMonitored1467290616
    test_log_scope.go:76: use -show-logs to present logs inline
    test_server_shim.go:152: automatically injected an external process virtual cluster under test; see comment at top of test_server_shim.go for details.
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=20</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBuildIsMemoryMonitored.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-46687

pkg/sql/schemachanger/scbuild/scbuild_test.TestBuildIsMemoryMonitored [failed](https://tanzanite.cluster.engflow.com/invocations/default/29d1f443-1dd0-47e5-810d-9b0a3c6e1535?testReportRun=10&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvc3FsL3NjaGVtYWNoYW5nZXIvc2NidWlsZDpzY2J1aWxkX3Rlc3Q=) on release-25.1 @ [0fb79d22b0fc0719f28a1e3993be2c4babf5155e](https://github.com/cockroachdb/cockroach/commits/0fb79d22b0fc0719f28a1e3993be2c4babf5155e):

Fatal error:

```
panic: test timed out after 15m0s
running tests:
	TestBuildIsMemoryMonitored (10m36s)

```
Stack: 

```
goroutine 277060 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x385
created by time.goFunc
	GOROOT/src/time/sleep.go:177 +0x2d
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestBuildIsMemoryMonitored
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestBuildIsMemoryMonitored1646586746
    test_log_scope.go:76: use -show-logs to present logs inline
    test_server_shim.go:152: automatically injected an external process virtual cluster under test; see comment at top of test_server_shim.go for details.
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=10</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBuildIsMemoryMonitored.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/sql/catalog/lease/lease.go
+++ b/pkg/sql/catalog/lease/lease.go
@@ -212,37 +212,42 @@ func (m *Manager) WaitForNoVersion(
 	return nil
 }
 
-// maybeGetDescriptorWithoutValidation gets a descriptor without validating
+// maybeGetDescriptorsWithoutValidation gets a descriptor without validating
 // from the KV layer.
-func (m *Manager) maybeGetDescriptorWithoutValidation(
-	ctx context.Context, id descpb.ID,
-) (desc catalog.Descriptor, err error) {
+func (m *Manager) maybeGetDescriptorsWithoutValidation(
+	ctx context.Context, ids descpb.IDs, existenceExpected bool,
+) (descs catalog.Descriptors, err error) {
 	err = m.storage.db.KV().Txn(ctx, func(ctx context.Context, txn *kv.Txn) (err error) {
 		const isDescriptorRequired = false
 		cr := m.storage.newCatalogReader(ctx)
-		c, err := cr.GetByIDs(ctx, txn, []descpb.ID{id}, isDescriptorRequired, catalog.Any)
+		c, err := cr.GetByIDs(ctx, txn, ids, isDescriptorRequired, catalog.Any)
 		if err != nil {
 			return err
 		}
-		desc = c.LookupDescriptor(id)
-		if desc == nil {
-			// Descriptor was dropped on us, so return a structured error. If
-			// WaitForInitialVersion is the caller then this indicates the descriptor
-			// was dropped right after creation.
-			return errors.Wrapf(catalog.ErrDescriptorNotFound, "descriptor %d could not be fetched to count leases", id)
+		descs = make(catalog.Descriptors, 0, len(ids))
+		for _, id := range ids {
+			desc := c.LookupDescriptor(id)
+			if desc == nil {
+				// Descriptor was dropped on us, so return a structured error.
+				if existenceExpected {
+					return errors.Wrapf(catalog.ErrDescriptorNotFound, "descriptor %d could not be fetched to count leases", id)
+				}
+			} else {
+				descs = append(descs, desc)
+			}
 		}
 		return nil
 	})
-	return desc, err
+	return descs, err
 }
 
 // countDescriptorsHeldBySessionIDs can be used to make sure certain nodes
-// (sessions) observe the existence of a given descriptor. Assuming the given
+// (sessions) observe the existence of a given set of descriptors. Assuming the given
 // sessions are still alive.
 func countDescriptorsHeldBySessionIDs(
 	ctx context.Context,
 	txn isql.Txn,
-	descID descpb.ID,
+	descIDs descpb.IDs,
 	region string,
 	sessionIDs []sqlliveness.SessionID,
 ) (int, error) {
@@ -257,17 +262,24 @@ func countDescriptorsHeldBySessionIDs(
 		}
 		b.WriteString(fmt.Sprintf("x'%s'", sessionID.String()))
 	}
+	d := strings.Builder{}
+	for _, descID := range descIDs {
+		if d.Len() > 0 {
+			d.WriteString(",")
+		}
+		d.WriteString(fmt.Sprintf("%d", descID))
+	}
 	// Query the count from the region.
 	row, err := txn.QueryRow(ctx, "wait-for-new-descriptor", txn.KV(),
 		fmt.Sprintf(
 			`
 SELECT count(*)
   FROM system.lease
- WHERE desc_id = %d
+ WHERE desc_id IN (%s)
        AND session_id IN (%s)
        AND crdb_internal.sql_liveness_is_alive(session_id)
        %s;`,
-			descID,
+			d.String(),
 			b.String(),
 			regionClause,
 		),
@@ -310,7 +322,7 @@ SELECT DISTINCT session_id FROM system.lease WHERE desc_id=%d AND crdb_internal.
 // (i.e. the optimizer memo will use the generation value as short circuit).
 func (m *Manager) WaitForInitialVersion(
 	ctx context.Context,
-	id descpb.ID,
+	descriptorsIds descpb.IDs,
 	retryOpts retry.Options,
 	regions regionliveness.CachedDatabaseRegions,
 ) error {
@@ -322,13 +334,15 @@ func (m *Manager) WaitForInitialVersion(
 	defer wsTracker.end()
 	decrAfterWait := m.IncGaugeAfterLeaseDuration(GaugeWaitForInitialVersion)
 	defer decrAfterWait()
+	// Track the set of descriptors, this will have descriptors removed on
+	// partial fulfillment (i.e. some subset having the initial version set).
+	var ids catalog.DescriptorIDSet
+	for _, id := range descriptorsIds {
+		ids.Add(id)
+	}
 	for lastCount, r := 0, retry.Start(retryOpts); r.Next(); {
-		desc, err := m.maybeGetDescriptorWithoutValidation(ctx, id)
+		descs, err := m.maybeGetDescriptorsWithoutValidation(ctx, ids.Ordered(), false /* existenceExpected */)
 		if err != nil {
-			if errors.Is(err, catalog.ErrDescriptorNotFound) {
-				log.Infof(ctx, "descriptor %d was dropped before initial leases was acquired", id)
-				return nil
-			}
 			return err
 		}
 		// If the object no longer exists or isn't fully created then skip this
@@ -339,134 +353,157 @@ func (m *Manager) WaitForInitialVersion(
 		// We don't need to worry about functions because their signature is stored
 		// inside the schema descriptor, which will cause an implicit bump invalidating
 		// cached metadata.
-		if desc == nil ||
-			(desc.DescriptorType() != catalog.Table && desc.DescriptorType() != catalog.Type) ||
-			desc.Dropped() ||
-			desc.Adding() {
+		descsToProcess := make([]catalog.Descriptor, 0, len(descs))
+		idsPerSchema := make(map[descpb.ID]descpb.IDs)
+
+		for _, desc := range descs {
+			if (desc.DescriptorType() != catalog.Table && desc.DescriptorType() != catalog.Type) ||
+				desc.Dropped() ||
+				desc.Adding() {
+				continue
+			}
+			descsToProcess = append(descsToProcess, desc)
+			idsPerSchema[desc.GetParentSchemaID()] = append(idsPerSchema[desc.GetParentSchemaID()], desc.GetID())
+		}
+		// No schemas to wait for.
+		if len(idsPerSchema) == 0 {
 			return nil
 		}
-		// Check to see if there are any leases that still exist on the previous
-		// version of the descriptor.
-		now := m.storage.clock.Now()
-		var count int
-		db := m.storage.db
-		// Get a list of sessions that had the schema leased out when this descriptor
-		// was created / modified.
-		var sessionsPerRegion map[string][]sqlliveness.SessionID
-		expectedSessions := 0
-		if err := db.Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
-			txn.KV().SetDebugName("wait-for-initial-lease-count-schema-leases")
-			// Look at what was leasing the schema at them modification time, we expect
-			// this be before the GC TTL because:
-			// 1) The defaults settings on the system database are not aggressive
-			// 2) We expect this to be a short wait in the recent past.
-			// If for some reason this timestamp is outside the GC TTL the transaction
-			// will get an error, which is a potential caveat here.
-			if err := txn.KV().SetFixedTimestamp(ctx, desc.GetModificationTime()); err != nil {
-				return err
-			}
-			expectedSessions = 0
-			sessionsPerRegion = make(map[string][]sqlliveness.SessionID)
+		// Go over each schema and set of descriptor IDs
+		totalCount := 0
+		totalExpectedCount := 0
+		for schemaID, descIDsForSchema := range idsPerSchema {
+			// Check to see if there are any leases that still exist on the previous
+			// version of the descriptor.
+			now := m.storage.clock.Now()
+			var count int
+			db := m.storage.db
+			// Get a list of sessions that had the schema leased out when this descriptor
+			// was created / modified.
+			var sessionsPerRegion map[string][]sqlliveness.SessionID
+			expectedSessions := 0
+			if err := db.Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
+				txn.KV().SetDebugName("wait-for-initial-lease-count-schema-leases")
+				// Look at what was leasing the schema at them modification time, we expect
+				// this be before the GC TTL because:
+				// 1) The defaults settings on the system database are not aggressive
+				// 2) We expect this to be a short wait in the recent past.
+				// If for some reason this timestamp is outside the GC TTL the transaction
+				// will get an error, which is a potential caveat here.
+				if err := txn.KV().SetFixedTimestamp(ctx, descsToProcess[0].GetModificationTime()); err != nil {
+					return err
+				}
+				expectedSessions = 0
+				sessionsPerRegion = make(map[string][]sqlliveness.SessionID)
 
-			prober := regionliveness.NewLivenessProber(db.KV(), m.storage.codec, regions, m.settings)
-			regionMap, err := prober.QueryLiveness(ctx, txn.KV())
-			if err != nil {
-				return err
-			}
-			// On single region clusters we can query everything at once.
-			if regionMap == nil {
-				sessionIDs, err := getSessionsHoldingDescriptor(ctx, txn, desc.GetParentSchemaID(), "")
+				prober := regionliveness.NewLivenessProber(db.KV(), m.storage.codec, regions, m.settings)
+				regionMap, err := prober.QueryLiveness(ctx, txn.KV())
 				if err != nil {
 					return err
 				}
-				sessionsPerRegion[""] = sessionIDs
-				expectedSessions += len(sessionIDs)
-			}
-			// Otherwise, process active schema leases by region, and use the
-			// region liveness subsystem to detect offline regions.
-			return regionMap.ForEach(func(region string) error {
-				var sessionIDs []sqlliveness.SessionID
-				var err error
-				if hasTimeout, timeout := prober.GetProbeTimeout(); hasTimeout {
-					err = timeutil.RunWithTimeout(ctx, "active-schema-leases-by-region", timeout, func(ctx context.Context) error {
-						var err error
-						sessionIDs, err = getSessionsHoldingDescriptor(ctx, txn, desc.GetParentSchemaID(), region)
+				// On single region clusters we can query everything at once.
+				if regionMap == nil {
+					sessionIDs, err := getSessionsHoldingDescriptor(ctx, txn, schemaID, "")
+					if err != nil {
 						return err
-					})
-				} else {
-					sessionIDs, err = getSessionsHoldingDescriptor(ctx, txn, desc.GetParentSchemaID(), region)
-				}
-				if err != nil {
-					return handleRegionLivenessErrors(ctx, prober, region, err)
-				}
-				sessionsPerRegion[region] = sessionIDs
-				expectedSessions += len(sessionIDs)
-				return nil
-			})
-		}); err != nil {
-			return err
-		}
-		// Next ensure the initial version exists on all nodes that have the schema
-		// leased out.
-		if err := db.Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
-			count = 0
-			txn.KV().SetDebugName("wait-for-initial-lease")
-			if err := txn.KV().SetFixedTimestamp(ctx, now); err != nil {
-				return err
-			}
-			prober := regionliveness.NewLivenessProber(db.KV(), m.storage.codec, regions, m.settings)
-			regionMap, err := prober.QueryLiveness(ctx, txn.KV())
-			if err != nil {
-				return err
-			}
-			// On multi-region we are going to process initial version on a per-region
-			// basis, which will allow us to query / inform region liveness of offline
-			// regions.
-			if regions != nil && regions.IsMultiRegion() {
-				return regionMap.ForEach(func(region string) error {
-					sessionIDs := sessionsPerRegion[region]
-					// Skip any regions without sessions.
-					if len(sessionIDs) == 0 {
-						return nil
 					}
-
-					var regionCount int
+					sessionsPerRegion[""] = sessionIDs
+					expectedSessions += len(sessionIDs)
+				}
+				// Otherwise, process active schema leases by region, and use the
+				// region liveness subsystem to detect offline regions.
+				return regionMap.ForEach(func(region string) error {
+					var sessionIDs []sqlliveness.SessionID
 					var err error
 					if hasTimeout, timeout := prober.GetProbeTimeout(); hasTimeout {
-						err = timeutil.RunWithTimeout(ctx, "wait-for-new-descriptor-by-region", timeout, func(ctx context.Context) error {
+						err = timeutil.RunWithTimeout(ctx, "active-schema-leases-by-region", timeout, func(ctx context.Context) error {
 							var err error
-							regionCount, err = countDescriptorsHeldBySessionIDs(ctx, txn, desc.GetID(), region, sessionIDs)
+							sessionIDs, err = getSessionsHoldingDescriptor(ctx, txn, schemaID, region)
 							return err
 						})
 					} else {
-						regionCount, err = countDescriptorsHeldBySessionIDs(ctx, txn, desc.GetID(), region, sessionIDs)
+						sessionIDs, err = getSessionsHoldingDescriptor(ctx, txn, schemaID, region)
 					}
-					if err := handleRegionLivenessErrors(ctx, prober, region, err); err != nil {
-						return err
+					if err != nil {
+						return handleRegionLivenessErrors(ctx, prober, region, err)
 					}
-					count += regionCount
+					sessionsPerRegion[region] = sessionIDs
+					expectedSessions += len(sessionIDs)
 					return nil
 				})
-			} else {
-				// Otherwise, we can query the initial versions directly.
-				count, err = countDescriptorsHeldBySessionIDs(ctx, txn, desc.GetID(), "", sessionsPerRegion[""])
+			}); err != nil {
 				return err
 			}
-		}); err != nil {
-			return err
+			// Next ensure the initial version exists on all nodes that have the schema
+			// leased out.
+			if err := db.Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
+				count = 0
+				txn.KV().SetDebugName("wait-for-initial-lease")
+				if err := txn.KV().SetFixedTimestamp(ctx, now); err != nil {
+					return err
+				}
+				prober := regionliveness.NewLivenessProber(db.KV(), m.storage.codec, regions, m.settings)
+				regionMap, err := prober.QueryLiveness(ctx, txn.KV())
+				if err != nil {
+					return err
+				}
+				// On multi-region we are going to process initial version on a per-region
+				// basis, which will allow us to query / inform region liveness of offline
+				// regions.
+				if regions != nil && regions.IsMultiRegion() {
+					return regionMap.ForEach(func(region string) error {
+						sessionIDs := sessionsPerRegion[region]
+						// Skip any regions without sessions.
+						if len(sessionIDs) == 0 {
+							return nil
+						}
+
+						var regionCount int
+						var err error
+						if hasTimeout, timeout := prober.GetProbeTimeout(); hasTimeout {
+							err = timeutil.RunWithTimeout(ctx, "wait-for-new-descriptor-by-region", timeout, func(ctx context.Context) error {
+								var err error
+								regionCount, err = countDescriptorsHeldBySessionIDs(ctx, txn, descIDsForSchema, region, sessionIDs)
+								return err
+							})
+						} else {
+							regionCount, err = countDescriptorsHeldBySessionIDs(ctx, txn, descIDsForSchema, region, sessionIDs)
+						}
+						if err := handleRegionLivenessErrors(ctx, prober, region, err); err != nil {
+							return err
+						}
+						count += regionCount
+						return nil
+					})
+				} else {
+					// Otherwise, we can query the initial versions directly.
+					count, err = countDescriptorsHeldBySessionIDs(ctx, txn, descIDsForSchema, "", sessionsPerRegion[""])
+					return err
+				}
+			}); err != nil {
+				return err
+			}
+			if count == expectedSessions*len(descIDsForSchema) {
+				// Remove descriptors that have had their wait completed.
+				for _, id := range descIDsForSchema {
+					ids.Remove(id)
+				}
+			}
+			totalCount += count
+			totalExpectedCount += expectedSessions * len(descIDsForSchema)
 		}
-		if count == expectedSessions {
+		// All the expected sessions are there now.
+		if totalCount == totalExpectedCount {
 			break
 		}
-		if count != lastCount {
-			log.Infof(ctx, "waiting for descriptor %d to appear on %d nodes. Last count was %d", desc.GetID(), expectedSessions, lastCount)
+		if totalCount != lastCount {
+			log.Infof(ctx, "waiting for descriptors %v to appear on %d nodes. Last count was %d", ids.Ordered(), totalExpectedCount, totalCount)
 			wsTracker.updateProgress(countDetail{
-				count:       count,
-				targetCount: expectedSessions,
+				count:       totalCount,
+				targetCount: totalExpectedCount,
 			})
 		}
-		lastCount = count
-
+		lastCount = totalCount
 	}
 	return nil
 }
@@ -496,9 +533,11 @@ func (m *Manager) WaitForOneVersion(
 	defer wsTracker.end()
 	for lastCount, r := 0, retry.Start(retryOpts); r.Next(); {
 		var err error
-		if desc, err = m.maybeGetDescriptorWithoutValidation(ctx, id); err != nil {
+		var descArr catalog.Descriptors
+		if descArr, err = m.maybeGetDescriptorsWithoutValidation(ctx, descpb.IDs{id}, true); err != nil {
 			return nil, err
 		}
+		desc = descArr[0]
 		// Check to see if there are any leases that still exist on the previous
 		// version of the descriptor.
 		now := m.storage.clock.Now()

--- a/pkg/sql/catalog/lease/lease_test.go
+++ b/pkg/sql/catalog/lease/lease_test.go
@@ -2580,7 +2580,7 @@ func TestLeaseWithOfflineTables(t *testing.T) {
 		// descriptor txn will not wait for the initial version.
 		if expected == descpb.DescriptorState_DROP && next == descpb.DescriptorState_PUBLIC {
 			require.NoError(t,
-				execCfg.LeaseManager.WaitForInitialVersion(ctx, testTableID(), retry.Options{}, nil))
+				execCfg.LeaseManager.WaitForInitialVersion(ctx, descpb.IDs{testTableID()}, retry.Options{}, nil))
 		}
 		// Wait for the lease manager's refresh worker to have processed the
 		// descriptor update.
@@ -4094,4 +4094,80 @@ func TestLongLeaseWaitMetrics(t *testing.T) {
 	require.Equal(t, int64(0), srv.MustGetSQLCounter("sql.leases.long_wait_for_no_version"))
 	require.Equal(t, int64(0), srv.MustGetSQLCounter("sql.leases.long_wait_for_two_version_invariant"))
 	require.Equal(t, int64(0), srv.MustGetSQLCounter("sql.leases.long_wait_for_one_version"))
+}
+
+// TestWaitForInitialVersionConcurrent this test is a basic sanity test that
+// intentionally has leases modified or used while the WaitForInitialVersion
+// is happening. The goal of this test is to find any concurrency related
+// bugs in the leasing subsystem with this logic.
+func TestWaitForInitialVersionConcurrent(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	ctx := context.Background()
+	s, sqlDB, _ := serverutils.StartServer(t, base.TestServerArgs{})
+	defer s.Stopper().Stop(ctx)
+	const numIter = 100
+	const numThreads = 4
+
+	expectedErrors := map[pgcode.Code]struct{}{
+		pgcode.UndefinedTable: {}, // table isn't created yet
+	}
+	tableCreator := func(ctx context.Context, index int) error {
+		stopTableUser := make(chan struct{})
+		defer close(stopTableUser)
+		tableUser := func(ctx context.Context, tblName string) error {
+			execLoop := true
+			// Execute a tight loop during the table creation
+			// process.
+			iter := 0
+			for execLoop {
+				select {
+				// Detect if we should move to the next
+				// table.
+				case <-stopTableUser:
+					execLoop = false
+				default:
+				}
+				var err error
+				// Execute DDL / DML we expect this to succeed or not notice the
+				// table.
+				if iter%2 == 0 {
+					_, err = sqlDB.Exec(fmt.Sprintf("ALTER TABLE %s ADD COLUMN j%d INT", tblName, iter))
+				} else {
+					_, err = sqlDB.Exec(fmt.Sprintf("SELECT * FROM %s", tblName))
+				}
+				// We expect failures when the table doesn't exist.
+				if err != nil {
+					var pqErr *pq.Error
+					if !errors.As(err, &pqErr) {
+						return err
+					}
+					if _, ok := expectedErrors[pgcode.MakeCode(string(pqErr.Code))]; !ok {
+						return err
+					}
+				}
+				iter += 1
+			}
+			return nil
+		}
+		tablePrefix := fmt.Sprintf("table%d_", index)
+		for i := 0; i < numIter; i++ {
+			tblName := fmt.Sprintf("%s%d", tablePrefix, i)
+			grp := ctxgroup.WithContext(ctx)
+			grp.GoCtx(func(ctx context.Context) error {
+				return tableUser(ctx, tblName)
+			})
+			_, err := sqlDB.Exec(fmt.Sprintf("CREATE TABLE %s(n int)", tblName))
+			if err != nil {
+				return err
+			}
+			stopTableUser <- struct{}{}
+			if err := grp.Wait(); err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+	require.NoError(t, ctxgroup.GroupWorkers(ctx, numThreads, tableCreator))
 }

--- a/pkg/sql/conn_executor_jobs.go
+++ b/pkg/sql/conn_executor_jobs.go
@@ -11,6 +11,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/jobs"
 	"github.com/cockroachdb/cockroach/pkg/jobs/jobspb"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
 	"github.com/cockroachdb/cockroach/pkg/sql/regions"
 	"github.com/cockroachdb/cockroach/pkg/util/retry"
 	"github.com/cockroachdb/errors"
@@ -61,19 +62,17 @@ func (ex *connExecutor) waitForInitialVersionForNewDescriptors(
 	// nodes that have leased the schema for them out are aware of the new object.
 	// This guarantees that any cached optimizer memos are discarded once the
 	// user transaction completes.
+	descriptorIDs := make(descpb.IDs, 0, len(ex.extraTxnState.descCollection.GetUncommittedTables()))
 	for _, tbl := range ex.extraTxnState.descCollection.GetUncommittedTables() {
 		if tbl.GetVersion() == 1 {
-			err := ex.planner.LeaseMgr().WaitForInitialVersion(ex.Ctx(), tbl.GetID(), retry.Options{
-				InitialBackoff: time.Millisecond,
-				MaxBackoff:     time.Second,
-				Multiplier:     1.5,
-			}, cachedRegions)
-			if err != nil {
-				return err
-			}
+			descriptorIDs = append(descriptorIDs, tbl.GetID())
 		}
 	}
-	return nil
+	return ex.planner.LeaseMgr().WaitForInitialVersion(ex.Ctx(), descriptorIDs, retry.Options{
+		InitialBackoff: time.Millisecond,
+		MaxBackoff:     time.Second,
+		Multiplier:     1.5,
+	}, cachedRegions)
 }
 
 // descIDsInSchemaChangeJobs returns all descriptor IDs with which schema change

